trin-core/src/portalnet/overlay.rs

```diff
@@ -0,0 +1,169 @@
+
+use discv5::enr::NodeId;
+use discv5::kbucket::{Filter, KBucketsTable};
+use parking_lot::RwLock;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Maximum number of ENRs in response to FindNodes.
+const FIND_NODES_MAX_NODES: usize = 32;
+/// Maximum numer of ENRs in response to FindContent.
+const FIND_CONTENT_MAX_NODES: usize = 32;
+
+#[derive(Clone)]
+pub struct Node {
+    enr: Enr,
+    data_radius: U256,
+}
+
+impl Node {
+    pub fn enr(&self) -> Enr {
+        self.enr.clone()
+    }
+
+    pub fn data_radius(&self) -> U256 {
+        self.data_radius.clone()
+    }
+}
+
+impl std::cmp::Eq for Node {}
+
+impl PartialEq for Node {
+    fn eq(&self, other: &Self) -> bool {
+        self.enr == other.enr
+    }
+}
+
+/// Configuration parameters for the overlay network.
+#[derive(Clone)]
+pub struct Config {
+    bucket_pending_timeout: Duration,
+    max_incoming_per_bucket: usize,
+    table_filter: Option<Box<dyn Filter<Node>>>,
+    bucket_filter: Option<Box<dyn Filter<Node>>>,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            bucket_pending_timeout: Duration::from_secs(60),
+            max_incoming_per_bucket: 16,
+            table_filter: None,
+            bucket_filter: None,
+        }
+    }
+}
+
+/// The node state for a node in an overlay network on top of Discovery v5.
+pub struct Overlay {
+    // The ENR of the local node.
+    local_enr: Arc<RwLock<Enr>>,
+    // The data radius of the local node.
+    data_radius: Arc<RwLock<U256>>,
+    // The routing table of the local node.
+    kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+}
+
+impl Overlay {
+    pub fn new(local_enr: Enr, data_radius: U256, config: Config) -> Self {
+        let kbuckets = Arc::new(RwLock::new(KBucketsTable::new(
+            local_enr.node_id().into(),
+            config.bucket_pending_timeout,
+            config.max_incoming_per_bucket,
+            config.table_filter,
+            config.bucket_filter,
+        )));
+        let local_enr = Arc::new(RwLock::new(local_enr));
+        let data_radius = Arc::new(RwLock::new(data_radius));
+
+        Self {
+            local_enr,
+            data_radius,
+            kbuckets,
+        }
+    }
+
+    /// Returns the local ENR of the node.
+    pub fn local_enr(&self) -> Enr {
+        self.local_enr.read().clone()
+    }
+
+    // Returns the data radius of the node.
+    pub fn data_radius(&self) -> U256 {
+        self.data_radius.read().clone()
+    }
+
+    /// Returns a vector of the ENRs of the closest nodes by the given distances.
+    pub fn nodes_by_distance(&self, mut distances: Vec<u64>) -> Vec<Enr> {
+        let mut nodes_to_send = Vec::new();
+        distances.sort_unstable();
+        distances.dedup();
+
+        if let Some(0) = distances.first() {
+            // if the distance is 0 send our local ENR
+            nodes_to_send.push(self.local_enr.read().clone());
+            distances.remove(0);
```

Comment:
Extremely minor nit, but removing the zeroth element would require shifting the whole `Vec`.

You could use a mutable binding:

```rust
let mut distances = distances.as_slice();
if let Some(0) = distances.first() {
    // TODO
    distances = &mut distances[1..];
}

Reply:
Doing this would require us to call `to_vec` on `distances` below when we call `nodes_by_distances` on `kbuckets`. If the memory overhead of the in-place shift versus the copy to a new vector are similar, then I think the existing code is easier to understand.

Reply:
Ugh, that [should be changed in `discv5` to a slice](https://github.com/sigp/discv5/blob/e3c7e308ef906dab57fb5088c0cf3c0a970475f4/src/kbucket.rs#L563). I can make a PR for that, unless you want to.

Reply:
Opened a PR [here](https://github.com/sigp/discv5/pull/86).